pkg/node/manager/manager.go

```diff
@@ -526,7 +527,11 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
 			}
 			oldNodeIPAddrs = append(oldNodeIPAddrs, address.IP)
 		}
+
 		m.deleteIPCache(oldNode.Source, oldNodeIPAddrs, ipsAdded)
+		for _, ip := range oldNodeIPAddrs {
+			m.removeFromIDMD(ip.String(), remoteHostIdentity, n.Source)
+		}
 
```

Comment:
I think you may need to filter this for loop's iteration to exclude `ipsAdded` from `oldNodeIPAddrs`. Otherwise, all old node IPs which still may be the same IPs in the "new" node are getting wiped from the metadata map.

Reply:
Ah yes, that was a bit of a naive attempt at the solution. I'm trying this as the solution:

```
diff --git a/pkg/node/manager/manager.go b/pkg/node/manager/manager.go
index 7ad1d94f4a13..4366d08f00f1 100644
--- a/pkg/node/manager/manager.go
+++ b/pkg/node/manager/manager.go
@@ -528,16 +528,13 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
                        oldNodeIPAddrs = append(oldNodeIPAddrs, address.IP)
                }
 
-               m.deleteIPCache(oldNode.Source, oldNodeIPAddrs, ipsAdded)
-               for _, ip := range oldNodeIPAddrs {
-                       m.removeFromIDMD(ip.String(), remoteHostIdentity, n.Source)
-               }
+               m.deleteIPCache(oldNode.Source, oldNodeIPAddrs, ipsAdded, remoteHostIdentity)
 
                // Delete the old health IP addresses if they have changed in this node.
-               m.deleteIPCache(oldNode.Source, []net.IP{oldNode.IPv4HealthIP, oldNode.IPv6HealthIP}, healthIPsAdded)
+               m.deleteIPCache(oldNode.Source, []net.IP{oldNode.IPv4HealthIP, oldNode.IPv6HealthIP}, healthIPsAdded, identity.IdentityUnknown)
 
                // Delete the old ingress IP addresses if they have changed in this node.
-               m.deleteIPCache(oldNode.Source, []net.IP{oldNode.IPv4IngressIP, oldNode.IPv6IngressIP}, ingressIPsAdded)
+               m.deleteIPCache(oldNode.Source, []net.IP{oldNode.IPv4IngressIP, oldNode.IPv6IngressIP}, ingressIPsAdded, identity.IdentityUnknown)
 
                entry.mutex.Unlock()
        } else {
@@ -580,7 +577,7 @@ func (m *Manager) removeFromIDMD(prefix string, id identity.NumericIdentity, src
 
 // deleteIPCache deletes the IP addresses from the IPCache with the 'oldSource'
 // if they are not found in the newIPs slice.
-func (m *Manager) deleteIPCache(oldSource source.Source, oldIPs []net.IP, newIPs []string) {
+func (m *Manager) deleteIPCache(oldSource source.Source, oldIPs []net.IP, newIPs []string, remoteID identity.NumericIdentity) {
        for _, address := range oldIPs {
                if address == nil {
                        continue
@@ -597,6 +594,9 @@ func (m *Manager) deleteIPCache(oldSource source.Source, oldIPs []net.IP, newIPs
                // added in this update.
                if !found {
                        m.ipcache.Delete(addrStr, oldSource)
+                       if remoteID != identity.IdentityUnknown {
+                               m.removeFromIDMD(addrStr, remoteID, oldSource)
+                       }
                }
        }
 }
```